test/src/lib/update-lock.spec.ts

```diff
@@ -0,0 +1,272 @@
+		mockFs({
+			[path.join(
+				constants.rootMountPoint,
+				constants.stateMountPoint,
+			)]: breadcrumbDir,
+		});
+	};
+
+	const mockLockfile = (appId: number, service: string) => {
+		mockFs({
+			[path.join(
+				constants.rootMountPoint,
+				updateLock.lockPath(appId),
+				service,
+			)]: {
+				'updates.lock': '',
+			},
+		});
+	};
+
+	// TODO: Remove these hooks when we don't need './test/data' as test process's rootMountPoint
+	before(() => {
+		// @ts-ignore // Set rootMountPoint for mockFs
+		constants.rootMountPoint = '/mnt/root';
+	});
+
+	after(() => {
+		// @ts-ignore
+		constants.rootMountPoint = process.env.ROOT_MOUNTPOINT;
+	});
+
+	describe('Lockfile path methods', () => {
+		const testAppId = 1234567;
+		const testService = 'test';
+
+		it('should return path prefix of service lockfiles on host', () => {
+			expect(updateLock.lockPath(testAppId)).to.equal(
+				`/tmp/balena-supervisor/services/${testAppId}`,
+			);
+			expect(updateLock.lockPath(testAppId, testService)).to.equal(
+				`/tmp/balena-supervisor/services/${testAppId}/${testService}`,
+			);
+		});
+
+		it('should return the complete paths of (non-)legacy lockfiles on host', () => {
+			const lockFilesOnHost = updateLock.__get__('lockFilesOnHost');
+			expect(lockFilesOnHost(testAppId, testService)).to.deep.equal([
+				`${constants.rootMountPoint}/tmp/balena-supervisor/services/${testAppId}/${testService}/updates.lock`,
+				`${constants.rootMountPoint}/tmp/balena-supervisor/services/${testAppId}/${testService}/resin-updates.lock`,
+			]);
+		});
+	});
+
+	describe('breadcrumbsExistOnHost', () => {
+		const breadcrumbsExistOnHost = updateLock.__get__('breadcrumbsExistOnHost');
+
+		it('should return whether breadcrumbs exist in the state partition on host', async () => {
+			// Should return true for single breadcrumb
+			for (const bc of breadcrumbFiles) {
+				mockBreadcrumbs(bc);
+				expect(await breadcrumbsExistOnHost()).to.be.true;
+				mockFs.restore();
+			}
+
+			// Should return true for multiple breadcrumbs
+			mockBreadcrumbs(...breadcrumbFiles);
+			expect(await breadcrumbsExistOnHost()).to.be.true;
+			mockFs.restore();
+		});
+	});
+
+	describe('Lock/dispose functionality', () => {
+		const lockFile = updateLock.__get__('lockFile');
+		const locksTaken = updateLock.__get__('locksTaken');
+		const dispose = updateLock.__get__('dispose');
+		const lockExistsErrHandler = updateLock.__get__('lockExistsErrHandler');
+
+		const releaseFn = stub();
+		const testLockPaths = ['/tmp/test/1', '/tmp/test/2'];
+
+		let unlockSyncStub: SinonStub;
+		let unlockAsyncSpy: SinonSpy;
+		let lockAsyncSpy: SinonSpy;
+
+		beforeEach(() => {
+			// @ts-ignore
+			unlockSyncStub = stub(lockFile, 'unlockSync').callsFake((lockPath) => {
+				// Throw error on process.exit for one of the two lockpaths
+				if (lockPath === testLockPaths[1]) {
+					throw new Error(
+						'handled unlockSync error which should not crash test process',
+					);
+				}
+			});
```

Comment:
This tests both error thrown + no error thrown case when attempting to cleanup lockfiles on Node process exit. The error thrown case should be handled silently by the code currently.